File: src/main/java/dev/slne/surf/essentials/commands/cheat/SpeedCommand.java
package dev.slne.surf.essentials.commands.cheat;

import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.slne.surf.essentials.commands.EssentialsCommand;
import dev.slne.surf.essentials.utils.EssentialsUtil;
import dev.slne.surf.essentials.utils.color.Colors;
import dev.slne.surf.essentials.utils.permission.Permissions;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.Range;

import java.util.Collection;
import java.util.List;

public class SpeedCommand extends EssentialsCommand {
    public SpeedCommand() {
        super("speed", "speed <speed> <walk | fly> [<players>]", "Change walk / fly speed", "flyspeed", "walkspeed");

        withRequirement(EssentialsUtil.checkPermissions(Permissions.SPEED_PERMISSION_SELF, Permissions.SPEED_PERMISSION_OTHER));

        then(floatArgument("speed", -10, 10)
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> detect(
                        sender.getCallee(),
                        List.of(getPlayerOrException(sender)),
                        args.getUnchecked("speed")
                ))
                .then(literal("walk")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeWalkSpeed(
                                sender.getCallee(),
                                List.of(getPlayerOrException(sender)),
                                args.getUnchecked("speed")
                        ))
                        .then(playersArgument("players")
                                .withPermission(Permissions.SPEED_PERMISSION_OTHER)
                                .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeWalkSpeed(
                                        sender.getCallee(),
                                        args.getUnchecked("players"),
                                        args.getUnchecked("speed")
                                ))))

                .then(literal("fly")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeFlySpeed(
                                sender.getCallee(),
                                List.of(getPlayerOrException(sender)),
                                args.getUnchecked("speed")
                        ))
                        .then(playersArgument("players")
                                .withPermission(Permissions.SPEED_PERMISSION_OTHER)
                                .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeFlySpeed(
                                        sender.getCallee(),
                                        args.getUnchecked("players"),
                                        args.getUnchecked("speed")
                                ))
                        )
                )
        )
                .then(literal("reset")
                        .executesNative((sender, args) -> changeWalkSpeed(
                                sender.getCallee(),
                                List.of(getPlayerOrException(sender)),
                                2.0f
                        ) + changeFlySpeed(
                                sender.getCallee(),
                                List.of(getPlayerOrException(sender)),
                                2.0f
                        ))
                        .then(playersArgument("players")
                                .withPermission(Permissions.SPEED_PERMISSION_OTHER)
                                .executesNative((sender, args) -> changeWalkSpeed(
                                        sender.getCallee(),
                                        args.getUnchecked("players"),
                                        2.0f
                                ) + changeFlySpeed(
                                        sender.getCallee(),
                                        args.getUnchecked("players"),
                                        2.0f
                                ))
                        )
                )

                .then(literal("resetwalk")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeWalkSpeed(
                                sender.getCallee(),
                                List.of(getPlayerOrException(sender)),
                                2.0f
                        ))
                        .then(playersArgument("players")
                                .withPermission(Permissions.SPEED_PERMISSION_OTHER)
                                .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeWalkSpeed(
                                        sender,
                                        args.getUnchecked("players"),
                                        2.0f
                                ))
                        )
                )

                .then(literal("resetfly")
                        .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeFlySpeed(
                                sender.getCallee(),
                                List.of(getPlayerOrException(sender)),
                                2.0f
                        ))
                        .then(playersArgument("players")
                                .withPermission(Permissions.SPEED_PERMISSION_OTHER)
                                .executesNative((NativeResultingCommandExecutor) (sender, args) -> changeFlySpeed(
                                        sender.getCallee(),
                                        args.getUnchecked("players"),
                                        2.0f
                                ))
                        )
                );
    }

    private int detect(CommandSender source, Collection<Player> playersUnchecked, @Range(from = -10, to = 10) Float speed) throws WrapperCommandSyntaxException {
        val players = EssentialsUtil.checkPlayerSuggestion(source, playersUnchecked);
        val calculatedSpeed = speed / 10f;
        int successes = 0;

        for (Player player : players) {
            if (player.isFlying()) {
                player.setFlySpeed(calculatedSpeed);
                successPlayer(player, "Fluggeschwindigkeit", speed);
            } else {
                player.setWalkSpeed(calculatedSpeed);
                successPlayer(player, "Gehgeschwindigkeit", speed);
            }
            successes++;
        }
        successSource(source, true, successes, speed, players);
        return successes;
    }

    private int changeFlySpeed(CommandSender source, Collection<Player> playersUnchecked, @Range(from = -10, to = 10) Float speed) throws WrapperCommandSyntaxException {
        return changeSpeed(source, EssentialsUtil.checkPlayerSuggestion(source, playersUnchecked), speed, false);
    }

    private int changeWalkSpeed(CommandSender source, Collection<Player> playersUnchecked, @Range(from = -10, to = 10) Float speed) throws WrapperCommandSyntaxException {
        return changeSpeed(source, EssentialsUtil.checkPlayerSuggestion(source, playersUnchecked), speed, true);
    }

    private int changeSpeed(CommandSender source, Collection<Player> players, @Range(from = -10, to = 10) float speed, boolean isWalkSpeed) throws WrapperCommandSyntaxException {
        val calculatedSpeed = speed / 10f;
        int successes = 0;

        for (Player player : players) {
            if (isWalkSpeed) {
                player.setWalkSpeed(calculatedSpeed);
                successPlayer(player, "Gehgeschwindigkeit", speed);
            } else {
                player.setFlySpeed(calculatedSpeed);
                successPlayer(player, "Fluggeschwindigkeit", speed);
            }
            successes++;
        }

        successSource(source, isWalkSpeed, successes, speed, players);
        return successes;
    }

    private void successPlayer(Player player, String mode, @Range(from = -10, to = 10) float speed) {
        EssentialsUtil.sendSuccess(player, Component.text("Deine %s wurde auf ".formatted(mode), Colors.SUCCESS)
                .append(Component.text(speed, Colors.TERTIARY))
                .append(Component.text(" gesetzt", Colors.SUCCESS)));
    }

    private void successSource(CommandSender source, boolean isWalkSpeed, int successes, @Range(from = -10, to = 10) float speed, Collection<Player> players) throws WrapperCommandSyntaxException {
        val mode = (isWalkSpeed) ? "Gehgeschwindigkeit" : "Fluggeschwindigkeit";
        if (successes == 1 && source instanceof Player player && player.getUniqueId() != players.iterator().next().getUniqueId()) {
            EssentialsUtil.sendSuccess(source, Component.text("Die %s von ".formatted(mode), Colors.SUCCESS)
                    .append(EssentialsUtil.getDisplayName(players.iterator().next()))
                    .append(Component.text(" wurde auf ", Colors.SUCCESS))
                    .append(Component.text(speed, Colors.TERTIARY))
                    .append(Component.text(" gesetzt", Colors.SUCCESS)));
        } else if (source instanceof Player player && player.getUniqueId() != players.iterator().next().getUniqueId()) {
            EssentialsUtil.sendSuccess(source, Component.text("Die %s von ".formatted(mode), Colors.SUCCESS)
                    .append(Component.text(successes, Colors.TERTIARY))
                    .append(Component.text("  Spielern wurde auf ", Colors.SUCCESS))
                    .append(Component.text(speed, Colors.TERTIARY))
                    .append(Component.text(" gesetzt", Colors.SUCCESS)));
        }
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/SpeedCommand.kt
package dev.slne.surf.essentials.command

import dev.jorel.commandapi.kotlindsl.*
import dev.slne.surf.essentials.util.EssentialsPermissionRegistry
import dev.slne.surf.surfapi.core.api.messages.adventure.sendText
import org.bukkit.entity.Player

private const val DEFAULT_SPEED_FLY = 0.1f
private const val DEFAULT_SPEED_WALK = 0.2f

fun speedCommand() = commandTree("speed") {
    withPermission(EssentialsPermissionRegistry.SPEED_COMMAND)
    playerExecutor { player, _ ->
        player.walkSpeed = DEFAULT_SPEED_WALK
        player.flySpeed = DEFAULT_SPEED_FLY

        player.sendText {
            appendPrefix()
            success("Deine Geh- und Fluggeschwindigkeit wurde zurückgesetzt.")
        }
    }

    literalArgument("walk") {
        withPermission(EssentialsPermissionRegistry.SPEED_COMMAND_WALK)
        playerExecutor { player, _ ->
            player.walkSpeed = DEFAULT_SPEED_WALK

            player.sendText {
                appendPrefix()
                success("Deine Gehgeschwindigkeit wurde zurückgesetzt.")
            }
        }
        doubleArgument("speed") {
            playerExecutor { player, arguments ->
                val speed: Double by arguments

                if (speed !in 0.0..10.0) {
                    player.sendText {
                        appendPrefix()
                        error("Die Geschwindigkeit muss zwischen 0 und 10 liegen.")
                    }
                    return@playerExecutor
                }

                player.walkSpeed = (speed.toFloat() / 10).coerceIn(0.0f, 1.0f)

                player.sendText {
                    appendPrefix()
                    success("Deine Gehgeschwindigkeit wurde auf ")
                    variableValue(speed.toString())
                    success(" gesetzt.")
                }
            }

            entitySelectorArgumentOnePlayer("player") {
                withPermission(EssentialsPermissionRegistry.SPEED_COMMAND_WALK_OTHERS)
                anyExecutor { executor, args ->
                    val player: Player by args
                    val speed: Double by args

                    if (speed !in 0.0..10.0) {
                        executor.sendText {
                            appendPrefix()
                            error("Die Geschwindigkeit muss zwischen 0 und 10 liegen.")
                        }
                        return@anyExecutor
                    }

                    player.walkSpeed = (speed.toFloat() / 10).coerceIn(0.0f, 1.0f)

                    executor.sendText {
                        appendPrefix()
                        success("Du hast die Gehgeschwindigkeit von ")
                        variableValue(player.name)
                        success(" auf ")
                        variableValue(speed.toString())
                        success(" gesetzt.")
                    }
                }
            }
        }
    }

    literalArgument("fly") {
        withPermission(EssentialsPermissionRegistry.SPEED_COMMAND_FLY)
        playerExecutor { player, _ ->
            player.flySpeed = DEFAULT_SPEED_FLY

            player.sendText {
                appendPrefix()
                success("Deine Fluggeschwindigkeit wurde zurückgesetzt.")
            }
        }
        doubleArgument("speed") {
            playerExecutor { player, arguments ->
                val speed: Double by arguments

                if (speed !in 0.0..10.0) {
                    player.sendText {
                        appendPrefix()
                        error("Die Geschwindigkeit muss zwischen 0 und 10 liegen.")
                    }
                    return@playerExecutor
                }

                player.flySpeed = (speed.toFloat() / 10).coerceIn(0.0f, 1.0f)

                player.sendText {
                    appendPrefix()
                    success("Deine Fluggeschwindigkeit wurde auf ")
                    variableValue(speed.toString())
                    success(" gesetzt.")
                }
            }

            entitySelectorArgumentOnePlayer("player") {
                withPermission(EssentialsPermissionRegistry.SPEED_COMMAND_FLY_OTHERS)
                anyExecutor { executor, args ->
                    val player: Player by args
                    val speed: Double by args

                    if (speed !in 0.0..10.0) {
                        executor.sendText {
                            appendPrefix()
                            error("Die Geschwindigkeit muss zwischen 0 und 10 liegen.")
                        }
                        return@anyExecutor
                    }

                    player.flySpeed = (speed.toFloat() / 10).coerceIn(0.0f, 1.0f)

                    executor.sendText {
                        appendPrefix()
                        success("Du hast die Fluggeschwindigkeit von ")
                        variableValue(player.name)
                        success(" auf ")
                        variableValue(speed.toString())
                        success(" gesetzt.")
                    }
                }
            }
        }
    }
}